tactics/Goalie_2.py

```diff
@@ -0,0 +1,510 @@
+		except rospy.ServiceException, e:
+			print("exception found")
+
+		if state:
+			state=state.stateB	
+	#     return state.ballVel.x >= 0 and state.ballPos.x>=0
+
+	def _protect_to_clear(self):
+		state = None
+		rospy.wait_for_service('bsServer',)
+		getState = rospy.ServiceProxy('bsServer',bsServer)
+		try:
+			state = getState(state)
+		except rospy.ServiceException, e:
+			print("exception found")
+
+		if state:
+			state=state.stateB	
+		return state.ballPos.x < -HALF_FIELD_MAXX + 2*OUR_GOAL_MAXX and state.ballVel.x < 0 and abs(state.ballPos.y) < OUR_GOAL_MAXY  
+
+
+	# note that execute_running() gets called BEFORE any of the execute_SUBSTATE methods gets called
+
+	def execute_peace(self):
+		print("Sab moh maya hain")
+		state = None
+
+		rospy.wait_for_service('bsServer',)
+		getState = rospy.ServiceProxy('bsServer',bsServer)
+		try:
+			state = getState(state)
+		except rospy.ServiceException, e:
+			print("exception found")
+
+		if state:
+			state=state.stateB	
+
+		if abs(state.ballPos.y) < OUR_GOAL_MAXY:
+			expected_y = state.ballPos.y
+		else:
+			if state.ballPos.y < 0:
+				expected_y = OUR_GOAL_MINY
+			else:
+				expected_y = OUR_GOAL_MAXY
+		print("In Peace :", state.ballPos.y)
+		point = Vector2D(-HALF_FIELD_MAXX + 5*BOT_RADIUS, expected_y)
+		start_time = rospy.Time.now()
+		start_time = 1.0*start_time.secs + 1.0*start_time.nsecs/pow(10,9)   
+		generatingfunction = _GoToPoint_.execute(start_time,BOT_BALL_THRESH)
+		for gf in generatingfunction:
+			self.kub,aim = gf
+			print("peace gf", point.x + point.y)
+			if not vicinity_points(aim,point,thresh=BOT_RADIUS) or self.outOfField(state):
+				self.behavior_failed = True
+				break
+
+
+		pass
+
+
+	def execute_clear(self):
+		print("Clear kar raha hoon")
+		target = Vector2D(4000,0)
+		state = None
+
+		rospy.wait_for_service('bsServer',)
+		getState = rospy.ServiceProxy('bsServer',bsServer)
+		try:
+			state = getState(state)
+		except rospy.ServiceException, e:
+			print("exception found")
+
+		if state:
+			state=state.stateB	
+		# self.gtB = GoToBall.GoToBall(BOT_BALL_THRESH)
+		# self.gtB.add_kub(self.kub)
+		# self.gtB.add_theta(theta=normalize_angle(atan2(target.y - state.ballPos.y,target.x - state.ballPos.x)))
+		# self.gtB.spin()
+
+		#Will Use KickToPoint after making it perfect with kicking
+
+
+		self.ktp = KickToPoint.KickToPoint(target)
+		self.ktp.add_kub(self.kub)
+		self.ktp.add_theta(theta=normalize_angle(atan2(target.y - state.ballPos.y,target.x - state.ballPos.x)))
+		self.ktp.spin()
+
+
+	def execute_active(self):
+		print("entered the active state")
+		state = None
+		rospy.wait_for_service('bsServer',)
+		getState = rospy.ServiceProxy('bsServer',bsServer)
+		try:
+			state = getState(state)
+		except rospy.ServiceException, e:
+			print("exception found")
+
+		if state:
+			state=state.stateB
+
+
+		if opponent_with_the_ball(state)!=None:
+			if state.ballVel.x < -10:
+				angle = math.atan(state.awayPos[state.opp_bot_closest_to_ball].theta)
```

Comment:
I think this is wrong. Can you explain why you are taking atan(bot_angle)?

Reply:
> I think this is wrong. Can you explain why you are taking atan(bot_angle)?

I used it from the previous version of goalie.I think it was when the ball was moving too fast so it faces the opponent bot and hence the ball. If you say should I replace it???

Reply:
See taking tan(bot_angle) and using that as slope is fine. But you are taking tan inverse of bot angle and then using it as slope of line